### PR TITLE
Variance

### PR DESCRIPTION
**Pull request by benjamin-hodgson**

### CODE REVIEW

File: drafts/2018-12-01-the-fourth-type-of-variance.md [hunk 1]
```diff
@@ -0,0 +1,204 @@
+---
+title: The Fourth Type of Variance
+---
+
+_Variance_ is a way to describe a relationship between different instantiations of the same polymorphic type. Many programmers know about covariance, contravariance, and invariance. Not so many know about the fourth type of variance, or the relationship between them.
```
Comment (laura-campbell) [inline drafts/2018-12-01-the-fourth-type-of-variance.md]: I would name what the fourth type of variance is here, so it says 
"not so many know about the fourth type of variance, ________, or the relationship between them"

File: drafts/2018-12-01-the-fourth-type-of-variance.md [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+
+Covariance
+----------
+
+Haskellers'll be familiar with _covariant functors_. It's the type of functor exhibited the the standard `Functor` class.
```
Comment (laura-campbell) [inline drafts/2018-12-01-the-fourth-type-of-variance.md]: word "the" duplicated

Comment (laura-campbell) [reply drafts/2018-12-01-the-fourth-type-of-variance.md]: i don't know what a functor is

Comment (laura-campbell) [reply drafts/2018-12-01-the-fourth-type-of-variance.md]: Maybe start each of these sections with a brief one liner describing that type of variance?

File: drafts/2018-12-01-the-fourth-type-of-variance.md [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+```haskell
+class Functor f where
+    fmap :: (a -> b) -> f a -> f b
+```
+
+This chimes with the intuition that a functor is a container of sorts. If you can turn each `a` in the container into a `b`, then you can turn a container of `a`s into a container of `b`s by converting each item in the container. The arrows go in the same direction.
```
Comment (laura-campbell) [inline drafts/2018-12-01-the-fourth-type-of-variance.md]: Not sure exactly what "the arrows go in the same direction" means here

File: drafts/2018-12-01-the-fourth-type-of-variance.md [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+A similar idea is visible in Scala's lattice of subtypes. Type parameters annotated with a `+` are covariant.
+
+```scala
+sealed abstract class List[+A]
+```
+
+A list of `Cat`s is a list of `Animal`s, because every `Cat` is an `Animal`. The subtype relationship of the container goes in the same direction as the subtype relationship of the elements. (In C# `+` is pronounced `out`, as in `IEnumerable<out T>`.)
```
Comment (laura-campbell) [inline drafts/2018-12-01-the-fourth-type-of-variance.md]: i'm confused by what the phrase " the subtype relationship of the container" means

File: drafts/2018-12-01-the-fourth-type-of-variance.md [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+```scala
+val cats : List[Cat] = List(Cat("Tilly"), Cat("Charlie"), Cat("Izzy"))
+val animals : List[Animal] = cats
+```
+
+Covariant things are producers.
```
Comment (laura-campbell) [inline drafts/2018-12-01-the-fourth-type-of-variance.md]: maybe elaborate on this phrase "covariant things are producers" - depends on your audience but i personally do not understand what this sentence means in this context

File: drafts/2018-12-01-the-fourth-type-of-variance.md [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+--------------
+
+Less familiar are _contravariant functors_:
+
+```haskell
+class Contravariant f where
+    contramap :: (b -> a) -> f a -> f b
+```
+
+The arrows go in opposite directions. A contravariant functor is a _consumer_ of `a`s. If you can turn `b`s into `a`s, then you can turn a consumer of `a`s into a consumer of `b`s by converting the `b`s into `a`s before they go into the consumer.
```
Comment (laura-campbell) [inline drafts/2018-12-01-the-fourth-type-of-variance.md]: i'm confused how this functor is a consumer of a's -- isn't it taking in Bs and turning them into As so it is producing As and consuming Bs? Once again, this could just be a "me" thing as I'm not familiar with these concepts.

**Review by laura-campbell — COMMENTED** [review]
I'm not sure any of my comments are helpful since this is so out of my depth, but in general I think explaining/defining terms/concepts (i.e. producer/consumer)  you are using before using them would be helpful for a junior dev! Once again, I'm assuming most people reading this will already know what all of these things are that I don't know, but it was all kind of over my head. Not sure whether it is my lack of computer science knowledge or lack of haskell/scala knowledge or just lack of general programming knowledge 😑 Maybe having someone not as super junior as me read it would be helpful, cause I was pretty lost 😭

**Comment by SaraJo** [comment]
Hi! I'm going to leave general feedback here. @laura-campbell noted the stuff that is Haskell specific (or maybe not, I just haven't worked with Functors before). 

I find a, b, c, type variable definitions really hard to follow. I know they are pretty, but they drive me nuts as they are not descriptive and it leaves me always with the question "WHY?". 

For example, in the section that says "This implies that the functor both consumes and produces as: you map items on the way out and on the way in." I've seen this in our JavaScripts (I need to figure out if we're using a minifier that is changing our variable names), when I see variable names like a, b, c, it makes it harder for me to know WHY we are doing something. 

I found this post to be great, super informative, I loved the drawings and I want to learn more.

**Comment by benjamin-hodgson** [comment]
@SaraJo and @laura-campbell, I restructured the post to spend quite a bit more time on the explanation of the three common types of variance, take another look and see if it's any easier to understand?
